<a href="https://colab.research.google.com/github/HunterAlpha7/Juncao/blob/main/ELP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment 2: Single Layer Perceptron

## Objective
Build a Single Layer Perceptron using only Python basics to predict whether a person is Fit (1) or Unfit (0) based on lifestyle indicators.

## Problem Statement
Predict Fit/Unfit based on:
1. Hours of physical activity per week
2. Average sleep hours per day
3. Daily water intake (liters)


## Task 1: Dataset Creation

# Labeling Rule:
# A person is considered Fit (1) if:
# - Physical activity >= 4 hours per week AND
# - Sleep >= 7 hours per day AND
# - Water intake >= 2.5 liters per day
# Otherwise, Unfit (0).

# This rule is based on general health guidelines (e.g., above-average activity, sufficient sleep, adequate hydration).
# Dataset is balanced with 8 Fit and 7 Unfit examples, synthesized from realistic averages (CDC, Mayo Clinic).

In [1]:
features = [
    [2.0, 6.5, 2.0],
    [5.0, 7.5, 3.0],
    [1.5, 5.8, 1.8],
    [6.0, 8.0, 2.8],
    [3.0, 6.0, 2.2],
    [4.5, 7.2, 3.2],
    [0.5, 7.0, 2.0],
    [7.0, 8.5, 3.5],
    [2.5, 6.2, 3.0],
    [5.5, 7.8, 2.7],
    [3.5, 5.5, 1.5],
    [8.0, 7.0, 3.8],
    [1.0, 8.0, 2.3],
    [4.0, 7.0, 2.5],
    [6.5, 7.5, 3.0]
]

labels = [0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1]

print("Dataset created with 15 points.")

Dataset created with 15 points.


## Task 2: Perceptron Initialization

# Weights initialized to 0 (fixed, to start from scratch without bias).
# Bias initialized to 0.
# Learning rate = 0.01 (chosen as small value to allow gradual convergence without overshooting, suitable for input scales).

In [2]:
weights = [0.0, 0.0, 0.0]
bias = 0.0
learning_rate = 0.01

## Task 3: Activation Function

# Computes weighted sum and applies step function.
# Threshold = 0 (standard for binary perceptron classification).

In [3]:
def activate(inputs, weights, bias):
    sum_ = bias
    for i in range(3):
        sum_ += inputs[i] * weights[i]
    return 1 if sum_ >= 0 else 0

## Task 4 & 5: Training Loop and Monitoring

# Update rule: For each misclassification,
# weight[j] += learning_rate * error * input[j]
# bias += learning_rate * error
# This is the standard perceptron update to move the decision boundary.
# Train for 100 epochs (sufficient to observe convergence).

In [4]:
epochs = 100
for epoch in range(epochs):
    total_error = 0
    for i in range(len(features)):
        pred = activate(features[i], weights, bias)
        error = labels[i] - pred
        total_error += abs(error)
        for j in range(3):
            weights[j] += learning_rate * error * features[i][j]
        bias += learning_rate * error
    print(f"Epoch {epoch+1}, Total Error: {total_error}, Weights: {weights}")

Epoch 1, Total Error: 11, Weights: [0.12, 0.0050000000000000044, 0.016999999999999998]
Epoch 2, Total Error: 5, Weights: [0.185, -0.04299999999999998, 0.032]
Epoch 3, Total Error: 5, Weights: [0.19000000000000003, -0.08299999999999996, 0.024]
Epoch 4, Total Error: 5, Weights: [0.2, -0.10999999999999996, 0.016000000000000007]
Epoch 5, Total Error: 2, Weights: [0.20500000000000002, -0.09499999999999995, 0.02600000000000001]
Epoch 6, Total Error: 5, Weights: [0.21500000000000005, -0.12199999999999994, 0.01800000000000001]
Epoch 7, Total Error: 2, Weights: [0.22000000000000006, -0.10699999999999993, 0.02800000000000001]
Epoch 8, Total Error: 6, Weights: [0.2700000000000001, -0.06399999999999992, 0.04500000000000001]
Epoch 9, Total Error: 3, Weights: [0.25500000000000006, -0.11399999999999991, 0.03500000000000002]
Epoch 10, Total Error: 4, Weights: [0.2850000000000001, -0.09399999999999989, 0.040000000000000015]
Epoch 11, Total Error: 3, Weights: [0.2600000000000001, -0.13899999999999987, 0

## Bonus: Calculate Accuracy

In [5]:
correct = 0
for i in range(len(features)):
    pred = activate(features[i], weights, bias)
    if pred == labels[i]:
        correct += 1
accuracy = correct / len(features) * 100
print(f"Accuracy: {accuracy}%")

Accuracy: 93.33333333333333%


## Task 6: User Input Testing

In [6]:
activity = float(input("Enter hours of physical activity per week: "))
sleep = float(input("Enter average sleep hours per day: "))
water = float(input("Enter daily water intake (liters): "))

user_input = [activity, sleep, water]
pred = activate(user_input, weights, bias)
if pred == 1:
    print("This person is likely to be Fit based on the given lifestyle pattern.")
else:
    print("This person is likely to be Unfit based on the given lifestyle pattern.")

Enter hours of physical activity per week: 3
Enter average sleep hours per day: 6
Enter daily water intake (liters): 2
This person is likely to be Unfit based on the given lifestyle pattern.


## Short Report

1. **Dataset Design**: Created 15 realistic data points based on health statistics (e.g., activity 0-8 hours, sleep 5.5-8.5, water 1.5-3.8). Varied to cover fit and unfit scenarios, balanced (8 fit, 7 unfit).

2. **Labeling Rule**: Fit if activity >=4, sleep >=7, water >=2.5; else Unfit.

3. **Checking Learning**: Monitored total error per epoch; it decreased over time (from ~11 to low values like 2-3), indicating convergence. Final accuracy ~93%.

4. **Weight Changes**: Weights started at [0,0,0]. Activity weight increased positively (emphasizing importance), sleep became negative (penalizing low sleep?), water positive. They stabilized with small adjustments.